File: src-tauri/src/extract.rs
use crate::error::AppError;
use std::fs::{self, File};
use std::io::{self, Cursor, Read};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

/// Check if a file is an archive that needs extraction
pub fn is_archive(path: &Path) -> bool {
    let ext = path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_lowercase());

    matches!(ext.as_deref(), Some("zip") | Some("7z") | Some("rar"))
}

/// Extract an archive to a destination directory
/// Returns the list of extracted VPK files
pub fn extract_archive(archive_path: &Path, dest_dir: &Path) -> Result<Vec<PathBuf>, AppError> {
    let ext = archive_path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_lowercase());

    match ext.as_deref() {
        Some("zip") => extract_zip(archive_path, dest_dir),
        Some("7z") => extract_7z(archive_path, dest_dir),
        Some("rar") => extract_rar(archive_path, dest_dir),
        _ => Err(AppError::Settings(format!(
            "Unknown archive format: {:?}",
            archive_path.extension()
        ))),
    }
}

/// Extract a ZIP archive
fn extract_zip(archive_path: &Path, dest_dir: &Path) -> Result<Vec<PathBuf>, AppError> {
    let file = File::open(archive_path)?;
    let mut archive = zip::ZipArchive::new(file)
        .map_err(|e| AppError::Settings(format!("Failed to open ZIP: {}", e)))?;

    let mut extracted_vpks = Vec::new();

    for i in 0..archive.len() {
        let mut file = archive
            .by_index(i)
            .map_err(|e| AppError::Settings(format!("Failed to read ZIP entry: {}", e)))?;

        let outpath = match file.enclosed_name() {
            Some(path) => dest_dir.join(path),
            None => continue,
        };

        // Skip directories and non-VPK files
        if file.is_dir() {
            continue;
        }

        // Only extract VPK files
        let is_vpk = outpath
            .extension()
            .and_then(|e| e.to_str())
            .map(|e| e.to_lowercase() == "vpk")
            .unwrap_or(false);

        if !is_vpk {
            continue;
        }

        // Create parent directories if needed
        if let Some(parent) = outpath.parent() {
            if !parent.exists() {
                fs::create_dir_all(parent)?;
            }
        }

        // Handle nested directories - flatten to dest_dir
        let file_name = outpath.file_name().unwrap();
        let final_path = dest_dir.join(file_name);

        let mut outfile = File::create(&final_path)?;
        io::copy(&mut file, &mut outfile)?;

        extracted_vpks.push(final_path);
    }

    Ok(extracted_vpks)
}

/// Extract a 7z archive
fn extract_7z(archive_path: &Path, dest_dir: &Path) -> Result<Vec<PathBuf>, AppError> {
    if !dest_dir.exists() {
        fs::create_dir_all(dest_dir)?;
    }

    // Read the archive into a buffer
    let mut file = File::open(archive_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    // Create a cursor for the buffer (implements Read + Seek)
    let cursor = Cursor::new(buffer);

    // Use sevenz-rust to decompress
    let extract_result =
        sevenz_rust::decompress_with_extract_fn(cursor, dest_dir, |entry, reader, dest| {
        // Get the entry name
        let name = entry.name();

        // Only extract VPK files
        if !name.to_lowercase().ends_with(".vpk") {
            return Ok(true); // Skip non-VPK files
        }

        // Flatten directory structure - just use the filename
        let file_name = match Path::new(name).file_name() {
            Some(file_name) => file_name,
            None => return Ok(true),
        };
        let final_path = dest.join(file_name);

        // Create the file
        let mut outfile = File::create(&final_path)
            .map_err(|e| sevenz_rust::Error::io(e))?;

        // Copy contents
        io::copy(reader, &mut outfile)
            .map_err(|e| sevenz_rust::Error::io(e))?;

        Ok(true)
    })
    ;

    if let Err(err) = extract_result {
        let temp_dir = create_temp_dir("modmanager-7z")?;
        let mut last_err = Some(format!("{:?}", err));

        for tool in ["7z", "7za"] {
            let run_result = run_command(
                tool,
                &[
                    "x",
                    "-y",
                    &format!(
                        "-o{}",
                        temp_dir
                            .to_str()
                            .ok_or_else(|| AppError::Settings("Invalid temp path".to_string()))?
                    ),
                    archive_path
                        .to_str()
                        .ok_or_else(|| AppError::Settings("Invalid archive path".to_string()))?,
                ],
            );

            match run_result {
                Ok(()) => {
                    let extracted = collect_vpks(&temp_dir)?;
                    let copied = copy_vpks_to_dest(&extracted, dest_dir)?;
                    let _ = fs::remove_dir_all(&temp_dir);
                    return Ok(copied);
                }
                Err(err) => last_err = Some(err),
            }
        }

        let message = if let Some(err) = last_err {
            format!(
                "Failed to extract 7z. Install '7z' (p7zip) or '7za' and try again. Details: {}",
                err
            )
        } else {
            "Failed to extract 7z. Install '7z' (p7zip) or '7za' and try again.".to_string()
        };

        let _ = fs::remove_dir_all(&temp_dir);
        return Err(AppError::Settings(message));
    }

    // Scan dest_dir for extracted VPK files
    let mut extracted_vpks = Vec::new();
    for entry in fs::read_dir(dest_dir)? {
        let entry = entry?;
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) == Some("vpk") {
            extracted_vpks.push(path);
        }
    }

    Ok(extracted_vpks)
}

/// Extract a RAR archive using available system tools (7z or unrar)
fn extract_rar(archive_path: &Path, dest_dir: &Path) -> Result<Vec<PathBuf>, AppError> {
    let temp_dir = create_temp_dir("modmanager-rar")?;

    let mut last_err = None;
    for tool in ["7z", "7za", "unrar"] {
        let result = match tool {
            "unrar" => run_command(
                tool,
                &[
                    "x",
                    "-y",
                    archive_path
                        .to_str()
                        .ok_or_else(|| AppError::Settings("Invalid archive path".to_string()))?,
                    temp_dir
                        .to_str()
                        .ok_or_else(|| AppError::Settings("Invalid temp path".to_string()))?,
                ],
            ),
            _ => run_command(
                tool,
                &[
                    "x",
                    "-y",
                    &format!(
                        "-o{}",
                        temp_dir
                            .to_str()
                            .ok_or_else(|| AppError::Settings("Invalid temp path".to_string()))?
                    ),
                    archive_path
                        .to_str()
                        .ok_or_else(|| AppError::Settings("Invalid archive path".to_string()))?,
                ],
            ),
        };

        match result {
            Ok(()) => {
                let extracted = collect_vpks(&temp_dir)?;
                let copied = copy_vpks_to_dest(&extracted, dest_dir)?;
                let _ = fs::remove_dir_all(&temp_dir);
                return Ok(copied);
            }
            Err(err) => last_err = Some(err),
        }
    }

    let message = if let Some(err) = last_err {
        format!(
            "RAR extraction failed. Install '7z' or 'unrar' and try again. Details: {}",
            err
        )
    } else {
        "RAR extraction failed. Install '7z' or 'unrar' and try again.".to_string()
    };

    let _ = fs::remove_dir_all(&temp_dir);
    Err(AppError::Settings(message))
}

fn run_command(cmd: &str, args: &[&str]) -> Result<(), String> {
    let output = Command::new(cmd).args(args).output();
    match output {
        Ok(output) if output.status.success() => Ok(()),
        Ok(output) => Err(format!(
            "{} exited with {}: {}",
            cmd,
            output.status,
            String::from_utf8_lossy(&output.stderr)
        )),
        Err(err) => Err(format!("{} failed to run: {}", cmd, err)),
    }
}

fn collect_vpks(root: &Path) -> Result<Vec<PathBuf>, AppError> {
    let mut vpks = Vec::new();
    collect_vpks_recursive(root, &mut vpks)?;
    Ok(vpks)
}

fn collect_vpks_recursive(root: &Path, out: &mut Vec<PathBuf>) -> Result<(), AppError> {
    for entry in fs::read_dir(root)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() {
            collect_vpks_recursive(&path, out)?;
            continue;
        }
        if path.extension().and_then(|e| e.to_str()) == Some("vpk") {
            out.push(path);
        }
    }
    Ok(())
}

fn copy_vpks_to_dest(
    extracted: &[PathBuf],
    dest_dir: &Path,
) -> Result<Vec<PathBuf>, AppError> {
    let mut copied = Vec::new();
    for path in extracted {
        let file_name = path
            .file_name()
            .ok_or_else(|| AppError::Settings("Invalid VPK filename".to_string()))?;
        let dest_path = dest_dir.join(file_name);
        fs::copy(path, &dest_path)?;
        copied.push(dest_path);
    }
    Ok(copied)
}

fn create_temp_dir(prefix: &str) -> Result<PathBuf, AppError> {
    let mut dir = std::env::temp_dir();
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|e| AppError::Settings(format!("Failed to create temp dir: {}", e)))?
        .as_nanos();
    dir.push(format!("{}-{}", prefix, nanos));
    fs::create_dir_all(&dir)?;
    Ok(dir)
}

/// Clean up archive file after successful extraction
pub fn cleanup_archive(archive_path: &Path) -> Result<(), AppError> {
    fs::remove_file(archive_path)?;
    Ok(())
}

File: electron/main/services/extract.ts
import { existsSync, mkdirSync, readdirSync, copyFileSync, unlinkSync } from 'fs';
import { join, extname, basename } from 'path';
import { randomBytes } from 'crypto';
import AdmZip from 'adm-zip';
import { spawn } from 'child_process';

/**
 * Find 7z executable path on Windows, checking common installation paths
 */
function find7zPath(): string[] {
    const candidates: string[] = [];

    // Check common Windows installation paths first
    const windowsPaths = [
        'C:\\Program Files\\7-Zip\\7z.exe',
        'C:\\Program Files (x86)\\7-Zip\\7z.exe',
    ];

    for (const p of windowsPaths) {
        if (existsSync(p)) {
            candidates.push(p);
        }
    }

    // Also try PATH-based commands as fallback
    candidates.push('7z', '7za');

    return candidates;
}

/**
 * Check if a file is an archive that needs extraction
 */
export function isArchive(filePath: string): boolean {
    const ext = extname(filePath).toLowerCase();
    return ext === '.zip' || ext === '.7z' || ext === '.rar';
}

/**
 * Extract an archive to a destination directory
 * Returns the list of extracted VPK files
 */
export async function extractArchive(
    archivePath: string,
    destDir: string
): Promise<string[]> {
    const ext = extname(archivePath).toLowerCase();

    switch (ext) {
        case '.zip':
            return extractZip(archivePath, destDir);
        case '.7z':
            return extract7z(archivePath, destDir);
        case '.rar':
            return extractRar(archivePath, destDir);
        default:
            throw new Error(`Unknown archive format: ${ext}`);
    }
}

/**
 * Extract a ZIP archive
 */
function extractZip(archivePath: string, destDir: string): string[] {
    const zip = new AdmZip(archivePath);
    const entries = zip.getEntries();
    const extractedVpks: string[] = [];

    for (const entry of entries) {
        if (entry.isDirectory) continue;

        const fileName = basename(entry.entryName);
        const ext = extname(fileName).toLowerCase();

        // Only extract VPK files
        if (ext !== '.vpk') continue;

        // Flatten to dest directory
        const destPath = join(destDir, fileName);
        zip.extractEntryTo(entry, destDir, false, true);
        extractedVpks.push(destPath);
    }

    return extractedVpks;
}

/**
 * Extract a 7z archive using system 7z command
 */
async function extract7z(archivePath: string, destDir: string): Promise<string[]> {
    // Create temp directory for extraction
    const tempDir = createTempDir('modmanager-7z');

    try {
        // Try common 7z paths (Windows install dirs + PATH fallback)
        for (const tool of find7zPath()) {
            try {
                await runCommand(tool, ['x', '-y', `-o${tempDir}`, archivePath]);
                const vpks = collectVpks(tempDir);
                const copied = copyVpksToDest(vpks, destDir);
                return copied;
            } catch {
                // Try next tool
            }
        }

        throw new Error(
            "Failed to extract 7z. Install '7z' (p7zip-full) or '7za' and try again."
        );
    } finally {
        // Cleanup temp directory
        try {
            rmDirRecursive(tempDir);
        } catch {
            // Ignore cleanup errors
        }
    }
}

/**
 * Extract a RAR archive using system unrar or 7z command
 */
async function extractRar(archivePath: string, destDir: string): Promise<string[]> {
    const tempDir = createTempDir('modmanager-rar');

    try {
        // Try common 7z paths, then unrar as fallback
        for (const tool of [...find7zPath(), 'unrar']) {
            try {
                if (tool === 'unrar') {
                    await runCommand(tool, ['x', '-y', archivePath, tempDir]);
                } else {
                    await runCommand(tool, ['x', '-y', `-o${tempDir}`, archivePath]);
                }
                const vpks = collectVpks(tempDir);
                const copied = copyVpksToDest(vpks, destDir);
                return copied;
            } catch {
                // Try next tool
            }
        }

        throw new Error(
            "RAR extraction failed. Install '7z' or 'unrar' and try again."
        );
    } finally {
        try {
            rmDirRecursive(tempDir);
        } catch {
            // Ignore cleanup errors
        }
    }
}

/**
 * Run a command and wait for it to complete
 * Includes timeout to prevent indefinite hangs (P1 fix #6)
 */
function runCommand(cmd: string, args: string[], timeoutMs = 300000): Promise<void> {
    return new Promise((resolve, reject) => {
        const proc = spawn(cmd, args, { stdio: 'pipe' });
        let stderr = '';
        let killed = false;

        // Set timeout to prevent indefinite hangs (5 minutes default)
        const timeoutId = setTimeout(() => {
            killed = true;
            proc.kill('SIGTERM');
            // Force kill after 5 seconds if still running
            setTimeout(() => {
                if (!proc.killed) {
                    proc.kill('SIGKILL');
                }
            }, 5000);
            reject(new Error(`${cmd} timed out after ${timeoutMs / 1000} seconds`));
        }, timeoutMs);

        proc.stderr?.on('data', (data) => {
            stderr += data.toString();
        });

        proc.on('close', (code) => {
            clearTimeout(timeoutId);
            if (killed) return; // Already rejected by timeout
            if (code === 0) {
                resolve();
            } else {
                reject(new Error(`${cmd} failed with code ${code}: ${stderr}`));
            }
        });

        proc.on('error', (err) => {
            clearTimeout(timeoutId);
            if (killed) return;
            reject(new Error(`${cmd} failed to run: ${err.message}`));
        });
    });
}

/**
 * Recursively collect VPK files from a directory
 */
function collectVpks(dir: string): string[] {
    const vpks: string[] = [];

    function walk(currentDir: string): void {
        if (!existsSync(currentDir)) return;

        const entries = readdirSync(currentDir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = join(currentDir, entry.name);
            if (entry.isDirectory()) {
                walk(fullPath);
            } else if (extname(entry.name).toLowerCase() === '.vpk') {
                vpks.push(fullPath);
            }
        }
    }

    walk(dir);
    return vpks;
}

/**
 * Copy VPK files to destination directory (flattening structure)
 */
function copyVpksToDest(vpks: string[], destDir: string): string[] {
    const copied: string[] = [];

    for (const vpk of vpks) {
        const fileName = basename(vpk);
        const destPath = join(destDir, fileName);
        copyFileSync(vpk, destPath);
        copied.push(destPath);
    }

    return copied;
}

/**
 * Create a temporary directory with cryptographically secure random name
 * (P0 security fix #3 - prevents race condition attacks)
 */
function createTempDir(prefix: string): string {
    const randomSuffix = randomBytes(16).toString('hex');
    const tmpDir = join(
        process.env.TMPDIR || process.env.TMP || '/tmp',
        `${prefix}-${randomSuffix}`
    );
    mkdirSync(tmpDir, { recursive: true, mode: 0o700 }); // Restrict permissions
    return tmpDir;
}

/**
 * Recursively remove a directory
 */
function rmDirRecursive(dir: string): void {
    if (!existsSync(dir)) return;

    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
        const fullPath = join(dir, entry.name);
        if (entry.isDirectory()) {
            rmDirRecursive(fullPath);
        } else {
            unlinkSync(fullPath);
        }
    }

    // Remove the directory itself
    const { rmdirSync } = require('fs');
    rmdirSync(dir);
}

/**
 * List contents of an archive (for Mina variants)
 */
export async function listArchiveContents(archivePath: string): Promise<string[]> {
    const ext = extname(archivePath).toLowerCase();

    if (ext === '.zip') {
        const zip = new AdmZip(archivePath);
        return zip.getEntries().map((e) => e.entryName);
    }

    // For 7z/rar, use 7z to list - try all candidates
    const candidates = find7zPath();

    const tryCandidate = (index: number): Promise<string[]> => {
        if (index >= candidates.length) {
            return Promise.reject(new Error('Failed to list archive contents. Install 7-Zip and try again.'));
        }

        return new Promise((resolve, reject) => {
            const proc = spawn(candidates[index], ['l', '-ba', archivePath], { stdio: 'pipe' });
            let stdout = '';

            proc.stdout?.on('data', (data) => {
                stdout += data.toString();
            });

            proc.on('close', (code) => {
                if (code === 0) {
                    // Parse 7z output - extract filenames
                    const lines = stdout.split('\n').filter((l) => l.trim());
                    const files = lines
                        .map((line) => {
                            // 7z -ba output format: date time attr size compressed name
                            const parts = line.trim().split(/\s+/);
                            return parts.slice(5).join(' ');
                        })
                        .filter((f) => f);
                    resolve(files);
                } else {
                    // Try next candidate
                    tryCandidate(index + 1).then(resolve).catch(reject);
                }
            });

            proc.on('error', () => {
                // Try next candidate
                tryCandidate(index + 1).then(resolve).catch(reject);
            });
        });
    };

    return tryCandidate(0);
}
